lib/Extension/LanguageServerPhpCsFixer/Provider/PhpCsFixerDiagnosticsProvider.php

```diff
@@ -0,0 +1,274 @@
+<?php
+
+namespace Phpactor\Extension\LanguageServerPhpCsFixer\Provider;
+
+use Amp\CancellationToken;
+use Amp\Promise;
+use Amp\Success;
+use Phpactor\Extension\LanguageServerPhpCsFixer\Model\PhpCsFixerProcess;
+use Phpactor\Extension\LanguageServerPhpCsFixer\Util\StringSharedChars;
+use Phpactor\LanguageServerProtocol\CodeAction;
+use Phpactor\LanguageServerProtocol\Command;
+use Phpactor\LanguageServerProtocol\Diagnostic;
+use Phpactor\LanguageServerProtocol\DiagnosticSeverity;
+use Phpactor\LanguageServerProtocol\Position;
+use Phpactor\LanguageServerProtocol\Range;
+use Phpactor\LanguageServerProtocol\TextDocumentItem;
+use Phpactor\LanguageServer\Core\CodeAction\CodeActionProvider;
+use Phpactor\LanguageServer\Core\Diagnostics\DiagnosticsProvider;
+use Psr\Log\LoggerInterface;
+use SebastianBergmann\Diff\Line;
+use SebastianBergmann\Diff\Parser;
+use LogicException;
+
+class PhpCsFixerDiagnosticsProvider implements DiagnosticsProvider, CodeActionProvider
+{
+    /** @var array<string, string> **/
+    private array $ruleDescriptions = [];
+
+    public function __construct(
+        private PhpCsFixerProcess $phpCsFixer,
+        private bool $showDiagnostics,
+        private LoggerInterface $logger,
+    ) {
+    }
+
+    /**
+     * @return Promise<Diagnostic[]>
+     */
+    public function provideDiagnostics(TextDocumentItem $textDocument, CancellationToken $cancel): Promise
+    {
+        if (!$this->showDiagnostics) {
+            return new Success([]);
+        }
+
+        return \Amp\call(function () use ($textDocument, $cancel) {
+            return yield $this->findDiagnostics($textDocument, $cancel);
+        });
+    }
+
+    public function provideActionsFor(TextDocumentItem $textDocument, Range $range, CancellationToken $cancel): Promise
+    {
+        return \Amp\call(function () use ($textDocument, $cancel) {
+            $diagnostics = yield $this->findDiagnostics($textDocument, $cancel);
+
+            if ($diagnostics === false) {
+                return [];
+            }
+
+            $title = 'Format with PHP CS Fixer';
+
+            $actions = [
+                CodeAction::fromArray([
+                    'title' => $title,
+                    'kind' => 'source.fixAll.phpactor.phpCsFixer',
+                    'diagnostics' => $diagnostics,
+                    'command' => new Command(
+                        $title,
+                        'php_cs_fixer.fix',
+                        [
+                            $textDocument->uri
+                        ]
+                    )
+                ])
+            ];
+
+            return $actions;
+        });
+    }
+
+    public function kinds(): array
+    {
+        return ['source.fixAll.phpactor.phpCsFixer'];
```

Comment:
Is there some standard here? AFAIK this is completely arbitrary, and existing things are `<refactor|quickfix>.<thing>`

Reply:
What I'm saying is this is an arbitrary identifier. AFAIK it has no semantic meaning? All the existing kinds are `refactor.blah`, or `quickfix.blah`. But maybe you know more than I do here?